misk-sqldelight/src/test/kotlin/misk/sqldelight/RetryingTransacterTest.kt

```diff
@@ -0,0 +1,161 @@
+    }
+
+    assertThat(retrievedTitle).isEqualTo(title)
+  }
+
+  @Test fun transactionIgnoresNonRetriableExceptions() {
+    transacter.transaction {
+      moviesDatabase.moviesQueries.createMovie(id, title)
+    }
+    var attempts = 0
+    assertThrows<NonRetriableException> {
+      transacter.transaction {
+        attempts += 1
+        moviesDatabase.moviesQueries.updateMovie(newTitle, Movies.Version(0), id)
+        throw NonRetriableException()
+      }
+    }
+    assertThat(attempts).isEqualTo(1)
+
+    val retrievedTitle = transacter.transactionWithResult {
+      moviesDatabase.moviesQueries.getMovie(id).executeAsOne().title
+    }
+
+    assertThat(retrievedTitle).isEqualTo(title)
+  }
+
+  @Test fun transactionWithResultRetriesAndSucceeds() {
+    transacter.transaction {
+      moviesDatabase.moviesQueries.createMovie(id, title)
+    }
+
+    var version = -2L
+
+    val updatedTitle = transacter.transactionWithResult {
+      version += 1
+      moviesDatabase.moviesQueries.updateMovie(newTitle, Movies.Version(version), id)
+      moviesDatabase.moviesQueries.getMovie(id).executeAsOne().title
+    }
+    assertThat(updatedTitle).isEqualTo(newTitle)
+
+    val retrievedTitle = transacter.transactionWithResult {
+      moviesDatabase.moviesQueries.getMovie(id).executeAsOne().title
+    }
+
+    assertThat(retrievedTitle).isEqualTo(newTitle)
+  }
+
+  @Test fun transactionWithResultExhaustsFailures() {
+    transacter.transaction {
+      moviesDatabase.moviesQueries.createMovie(id, title)
+    }
+
+    assertThrows<OptimisticLockException> {
+      transacter.transactionWithResult {
+        moviesDatabase.moviesQueries.updateMovie(newTitle, Movies.Version(-3L), id)
+      }
+    }
+
+
+    val retrievedTitle = transacter.transactionWithResult {
+      moviesDatabase.moviesQueries.getMovie(id).executeAsOne().title
+    }
+
+    assertThat(retrievedTitle).isEqualTo(title)
+  }
+
+  @Test fun transactionWithResultIgnoresNonRetriableExceptions() {
+    transacter.transaction {
+      moviesDatabase.moviesQueries.createMovie(id, title)
+    }
+    var attempts = 0
+    assertThrows<NonRetriableException> {
+      transacter.transactionWithResult {
+        attempts += 1
+        moviesDatabase.moviesQueries.updateMovie(newTitle, Movies.Version(0), id)
+        throw NonRetriableException()
+      }
+    }
+    assertThat(attempts).isEqualTo(1)
+
+    val retrievedTitle = transacter.transactionWithResult {
+      moviesDatabase.moviesQueries.getMovie(id).executeAsOne().title
+    }
+
+    assertThat(retrievedTitle).isEqualTo(title)
+  }
+
+  @Test fun nestedRetries() {
+    var tries = 0
+    val oneRetryTransacter = transacter.maxAttempts(2)
+    assertThrows<OptimisticLockException> {
+      oneRetryTransacter.transaction {
+        oneRetryTransacter.transaction {
+          tries++
+          throw OptimisticLockException("fake transient exception")
+        }
+      }
+    }
```

Comment:
I've done a crude implementation of 2 tracking the state on a threadlocal. The actual transacter internals from sqldelight _does_ track if we're in a transaction or not - but we don't have access to that in the scope of `RetryTransacter`

Reply:
I think this work for now. I worry this can have unintended consequences if folks are using coroutines during their db transaction, but let's keep this for now and iterate if it's needed.